hudi-cli/src/main/java/org/apache/hudi/cli/commands/MetadataCommand.java

```diff
@@ -0,0 +1,202 @@
+  }
+
+  @CliCommand(value = "metadata delete", help = "Remove the Metadata Table")
+  public String delete() throws Exception {
+    HoodieTableMetaClient metaClient = HoodieCLI.getTableMetaClient();
+    Path metadataPath = new Path(HoodieMetadata.getMetadataTableBasePath(HoodieCLI.basePath));
+    try {
+      FileStatus[] statuses = HoodieCLI.fs.listStatus(metadataPath);
+      if (statuses.length > 0) {
+        HoodieCLI.fs.delete(metadataPath, true);
+      }
+    } catch (FileNotFoundException e) {
+      // Metadata directory does not exist
+    }
+
+    HoodieMetadata.remove(HoodieCLI.basePath);
+
+    return String.format("Removed Metdata Table from %s", metadataPath);
+  }
+
+  @CliCommand(value = "metadata init", help = "Update the metadata table from commits since the creation")
+  public String init(@CliOption(key = {"readonly"}, unspecifiedDefaultValue = "false",
+      help = "Open in read-only mode") final boolean readOnly) throws Exception {
+    HoodieTableMetaClient metaClient = HoodieCLI.getTableMetaClient();
+    Path metadataPath = new Path(HoodieMetadata.getMetadataTableBasePath(HoodieCLI.basePath));
+    try {
+      FileStatus[] statuses = HoodieCLI.fs.listStatus(metadataPath);
+    } catch (FileNotFoundException e) {
+      // Metadata directory does not exist
+      throw new RuntimeException("Metadata directory (" + metadataPath.toString() + ") does not exist.");
+    }
+
+    HoodieMetadata.remove(HoodieCLI.basePath);
+
+    long t1 = System.currentTimeMillis();
+    if (readOnly) {
+      HoodieMetadata.init(HoodieCLI.conf, HoodieCLI.basePath);
+    } else {
+      HoodieWriteConfig writeConfig = HoodieWriteConfig.newBuilder().withPath(HoodieCLI.basePath)
+          .withUseFileListingMetadata(true).build();
+      initJavaSparkContext();
+      HoodieMetadata.init(jsc, writeConfig);
+    }
+    long t2 = System.currentTimeMillis();
+
+    String action = readOnly ? "Opened" : "Initialized";
+    return String.format(action + " Metadata Table in %s (duration=%.2fsec)", metadataPath, (t2 - t1) / 1000.0);
+  }
+
+  @CliCommand(value = "metadata stats", help = "Print stats about the metadata")
+  public String stats() throws IOException {
+    HoodieTableMetaClient metaClient = HoodieCLI.getTableMetaClient();
+    Map<String, String> stats = HoodieMetadata.getStats(HoodieCLI.basePath, true);
+
+    StringBuffer out = new StringBuffer("\n");
+    out.append(String.format("Base path: %s\n", HoodieMetadata.getMetadataTableBasePath(HoodieCLI.basePath)));
+    for (Map.Entry<String, String> entry : stats.entrySet()) {
+      out.append(String.format("%s: %s\n", entry.getKey(), entry.getValue()));
+    }
+
+    return out.toString();
+  }
+
+  @CliCommand(value = "metadata list-partitions", help = "Print a list of all partitions from the metadata")
+  public String listPartitions() throws IOException {
+    HoodieTableMetaClient metaClient = HoodieCLI.getTableMetaClient();
```

Comment:
Should there be call to     HoodieMetadata.init(HoodieCLI.conf, HoodieCLI.basePath) ?

Reply:
The MetaCommand code assumes that you will call "metadata init" first. This is to ensure that listPartitions can be performed both with and without the metadata being initialized for performance comparison purpose.

Reply:
> > I instrumented the code to track writeConfig.useFileListingMetadata. It looks like it is being disabled for a dataset that has the metadata dataset
> > 20/09/26 02:58:11 ERROR HoodieMetadata: writeConfig.useFileListingMetadata() =true
> > 20/09/26 02:58:14 ERROR HoodieMetadata: writeConfig.useFileListingMetadata() =false
> > The above line is coming from HoodieMetadata.init() method. I am unable to test using spark.write() as it seems to fallback to listing
> 
> I think the second log (writeConfig.useFileListingMetadata() =false) is coming from the Hoodie Metadata Table's internal writeConfig. Since the HUDI Metadata Table is also a HUDI Table, it also has a HoodieWriteConfig wherein the useFileListingMetadata is disabled.
> 
> Fallback to listing is by design to prevent issues during adoption.

Got it, So, the spark.write.xxxx() works ?

Reply:
> Adding a note as part of testing : I am not able to query the metadata table after performing a simple bulk-insert on a non-partitioned dataset.
> 

I tested the partition-less case on spark-shell. It does not work with current code. 

The issue seems to be that in the partition-less case, the name of the partition saved within the Metadata Table is "" (empty) so the FileStatus returned from the metadata table listings are not correct absolute path (first character of file name is missing). I will add a task to fix this case. 

@bvaradar Can you please test with a partitioned case?

Reply:
Fixed.